### PR TITLE
[Danger] Support showing the build times when something goes above a threshold

### PR DESCRIPTION
**Pull request by orta**
There's been a lot of talk in the community about ways to speed up builds, notably [thatthinginswift](https://thatthinginswift.com/debug-long-compile-times-swift/) and from [bryan irace](http://irace.me/swift-profiling). These offer one-off ways to check when you think something is slow. 

Realistically though, what happens is a long tail of small methods that add up and slowly consume your sanity - till it looks [like code Injection](http://artsy.github.io/blog/2016/03/05/iOS-Code-Injection/) is the only way to work productively.

Luckily we have [Danger](https://github.com/danger/danger/), this PR is adding the ability for us to find out in our CI whether we're introducing code that is taking a long time to compile - either from pods or our app.

Incase you were wondering, Eigen's swift build time wasn't terrible:

``` sh
~/d/r/Rainforest (master) ⏛  pbpaste | egrep '\.[0-9]ms' | sort -t "." -k 1 -n | tail -10
62.2ms    /Users/orta/dev/ios/apps/eigen/Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionToolbarView.swift:49:10    func setupUsingState(lotState: LotState)
71.1ms    /Users/orta/dev/ios/apps/eigen/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotSetViewController.swift:29:19    @objc override func viewDidLoad()
73.2ms    /Users/orta/dev/ios/apps/eigen/Artsy/Views/Auction/AuctionTitleView.swift:90:10    func titleView() -> UIView
79.0ms    /Users/orta/dev/ios/apps/eigen/Artsy/View_Controllers/Auction/AuctionViewController.swift:123:10    @objc dynamic func setupForSale(saleViewModel: SaleViewModel)
85.2ms    /Users/orta/dev/ios/apps/eigen/Artsy/View_Controllers/Auction/RefinementOptionsViewController+Private.swift:63:10    func stackView() -> ORStackView
86.8ms    /Users/orta/dev/ios/apps/eigen/Artsy/App/ARHockeyFeedbackDelegate.swift:19:10    @objc func showFeedback(image: UIImage? = default)
92.0ms    /Users/orta/dev/ios/apps/eigen/Artsy/Views/Auction/AuctionBannerView.swift:41:18    private func setupViews()
188.3ms    /Users/orta/dev/ios/apps/eigen/Pods/Starscream/Source/WebSocket.swift:528:18    private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
215.7ms    /Users/orta/dev/ios/apps/eigen/Pods/Starscream/Source/WebSocket.swift:528:18    private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
551.7ms    /Users/orta/dev/ios/apps/eigen/Artsy/View_Controllers/Auction/AuctionViewController.swift:176:10    @objc dynamic func showRefineTappedAnimated(animated: Bool)
```

But as [mad-eye moody](https://encrypted.google.com/search?hl=en&q=HPMOR%20constant%20vigilence#hl=en&q=HPMOR+constant+vigilance) says, Constant Vigilance.

### CODE REVIEW

**Comment by alloy** [comment]
What about reversing the sort? I keep expecting the worst offender to be at the top.

**Comment by orta** [comment]
Agreed

**Comment by orta** [comment]
Interesting note: running it on Circle was about 2x faster than on my rMBP

``` sh
~/d/r/Rainforest (master) ⏛  pbpaste | egrep '\.[0-9]ms' | sort -t "." -k 1 -n | tail -10
29.2ms	/Users/distiller/eigen/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift:50:19	@objc override func viewDidLoad()
30.8ms	/Users/distiller/eigen/Artsy/Views/Auction/AuctionTitleView.swift:90:10	func titleView() -> UIView
32.6ms	/Users/distiller/eigen/Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionLotImagesPreviewView.swift:29:5	init(progressSignal: Observable<CGFloat>, nextSignal: Observable<Int>, salesPerson: LiveAuctionsSalesPersonType)
33.2ms	/Users/distiller/eigen/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotSetViewController.swift:29:19	@objc override func viewDidLoad()
37.9ms	/Users/distiller/eigen/Artsy/App/ARHockeyFeedbackDelegate.swift:19:10	@objc func showFeedback(image: UIImage? = default)
39.9ms	/Users/distiller/eigen/Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionToolbarView.swift:49:10	func setupUsingState(lotState: LotState)
43.4ms	/Users/distiller/eigen/Artsy/Views/Auction/AuctionBannerView.swift:41:18	private func setupViews()
57.2ms	/Users/distiller/eigen/Artsy/View_Controllers/Auction/RefinementOptionsViewController+Private.swift:63:10	func stackView() -> ORStackView
170.9ms	/Users/distiller/eigen/Pods/Starscream/Source/WebSocket.swift:528:18	private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
246.2ms	/Users/distiller/eigen/Artsy/View_Controllers/Auction/AuctionViewController.swift:176:10	@objc dynamic func showRefineTappedAnimated(animated: Bool)
```

**Comment by alloy** [comment]
Hah. Ideally the danger threshold should probably not be a fixed number, but rather an algorithm that finds outliers based on the average, or something.

**Comment by orta** [comment]
That's cleverer than my UNIX skills @alloy. 😜

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">1 Warning</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td>PR is classed as Work in Progress</td>
    </tr>
    
  </tbody>
</table>
  

  


<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

**Comment by alloy** [comment]
Maybe one of these?

* http://stackoverflow.com/a/5892661/95397
* https://github.com/savanna-initiative/savanna-outliers

**Comment by orta** [comment]
Hrm, well, if we're going into ruby, that's going to add a bunch of code to our `Dangerfile`, let's see what I can do

**Comment by alloy** [comment]
You can probably do something with AWK, but that sounds like it would take a lot more time to figure out.

**Comment by orta** [comment]
screen shot 2016-05-03 at 14 01 49

Interesting, the top two will probably will show up pretty well as outliers

**Comment by orta** [comment]
When using 3 standard deviations we get these: 

screen shot 2016-05-03 at 14 15 46

**Comment by alloy** [comment]
Looks sweet to me 👌 

PS s/enumable/enumerable/

File: config/enumerable_stats.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    m = self.mean
+    sum = self.inject(0){|accum, i| accum +(i-m)**2 }
+    (1/self.length.to_f*sum)
+  end
+
+  def standard_deviation
+    return Math.sqrt(self.sample_variance)
+  end
+
+  def valuesOutsideStandardDeviation(deviation)
```
Comment (orta) [inline config/enumerable_stats.rb]: opinions on a better, more ruby-ish, name?

Comment (alloy) [inline config/enumerable_stats.rb]: `outliers(deviation)` ?

Comment (orta) [inline config/enumerable_stats.rb]: I'm alright with it, but it's not _too_ descriptive, though the language comes with `tr` so who am I to talk

**Comment by mennenia** [comment]
so where does it get the average from? Several builds, or the different actions in the files like the list you posted:

```
29.2ms  /Users/distiller/eigen/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift:50:19    @objc override func viewDidLoad()
30.8ms  /Users/distiller/eigen/Artsy/Views/Auction/AuctionTitleView.swift:90:10 func titleView() -> UIView
32.6ms  /Users/distiller/eigen/Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionLotImagesPreviewView.swift:29:5    init(progressSignal: Observable<CGFloat>, nextSignal: Observable<Int>, salesPerson: LiveAuctionsSalesPersonType)
33.2ms  /Users/distiller/eigen/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotSetViewController.swift:29:19 @objc override func viewDidLoad()
37.9ms  /Users/distiller/eigen/Artsy/App/ARHockeyFeedbackDelegate.swift:19:10   @objc func showFeedback(image: UIImage? = default)
39.9ms  /Users/distiller/eigen/Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionToolbarView.swift:49:10    func setupUsingState(lotState: LotState)
43.4ms  /Users/distiller/eigen/Artsy/Views/Auction/AuctionBannerView.swift:41:18    private func setupViews()
57.2ms  /Users/distiller/eigen/Artsy/View_Controllers/Auction/RefinementOptionsViewController+Private.swift:63:10   func stackView() -> ORStackView
170.9ms /Users/distiller/eigen/Pods/Starscream/Source/WebSocket.swift:528:18    private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
246.2ms /Users/distiller/eigen/Artsy/View_Controllers/Auction/AuctionViewController.swift:176:10    @objc dynamic func showRefineTappedAnimated(animated: Bool)
```

and checks for outliers in there?

**Comment by orta** [comment]
I *think* it's every class, closure and method being compiled by the swift compiler, so it takes the 1000 most expensive to compile then looks in them for outliers. No state between builds - that is asking for trouble.

**Comment by mennenia** [comment]
Thanks for the clarification! Makes sense and figured, was just curious.

**Comment by orta** [comment]
Sidenote: if you're wondering how to do pretty complex work with Danger, you can use `  env CIRCLE_ARTIFACTS="." bundle exec danger local` with a downloaded version of the [artefacts from circle](https://circleci.com/gh/artsy/eigen/2383#artifacts)

File: config/enumerable_stats.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  def sample_variance
+    m = self.mean
+    sum = self.inject(0){|accum, i| accum +(i-m)**2 }
+    (1/self.length.to_f*sum)
+  end
+
+  def standard_deviation
+    return Math.sqrt(self.sample_variance)
```
Comment (alloy) [inline config/enumerable_stats.rb]: `return`????

**Comment by alloy** [comment]
This is awesome work, @orta 👏

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
        
            3 Warnings
        
      </th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td data-sticky="true">PR is classed as Work in Progress</td>
    </tr>
        <tr>
      <td>:warning:</td>
      <td data-sticky="true">Needs testing on a Phone if change is non-trivial</td>
    </tr>
        <tr>
      <td>:warning:</td>
      <td data-sticky="true">Detected some Swift building time outliers</td>
    </tr>
    
    
  </tbody>
</table>
  

  



Time | Class | Function |
| --- | ----- | ----- |231.9ms | [AuctionViewController.swift](/artsy/eigen/tree/orta-danger_speed/Artsy/View_Controllers/Auction/AuctionViewController.swift#L176) | @objc dynamic func showRefineTappedAnimated(animated: Bool)
172.3ms | [WebSocket.swift](/artsy/eigen/tree/orta-danger_speed/Pods/Starscream/Source/WebSocket.swift#L528) | private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
57.8ms | [RefinementOptionsViewController+Private.swift](/artsy/eigen/tree/orta-danger_speed/Artsy/View_Controllers/Auction/RefinementOptionsViewController+Private.swift#L63) | func stackView() -> ORStackView
42.0ms | [AuctionBannerView.swift](/artsy/eigen/tree/orta-danger_speed/Artsy/Views/Auction/AuctionBannerView.swift#L41) | private func setupViews()
40.0ms | [LiveAuctionToolbarView.swift](/artsy/eigen/tree/orta-danger_speed/Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionToolbarView.swift#L49) | func setupUsingState(lotState: LotState)
37.7ms | [ARHockeyFeedbackDelegate.swift](/artsy/eigen/tree/orta-danger_speed/Artsy/App/ARHockeyFeedbackDelegate.swift#L19) | @objc func showFeedback(image: UIImage? = default)




<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

**Comment by orta** [comment]
I feel like something is off with the build timings. I wonder if it's doing incremental builds?

File: Dangerfile [hunk 3]
```diff
@@ -67,7 +67,7 @@ most_expensive_swift_table = `cat #{build_file} | egrep '\.[0-9]ms' | sort -t ".
 
 # each line looks like "29.2ms  /Users/distiller/eigen/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift:50:19    @objc override func viewDidLoad()"
 # Looks for outliers based on http://stackoverflow.com/questions/5892408/inferential-statistics-in-ruby/5892661#5892661
-time_values = most_expensive_swift_table.lines.map { |line| line.split.first.to_i }
+time_values = most_expensive_swift_table.lines.map { |line| line.split.first.to_i }.reject { |value| value == 0 }
```
Comment (alloy) [inline Dangerfile]: For future reference: `.reject(&:zero?)`

File: config/enumerable_stats.rb [hunk 4]
```diff
@@ -4,27 +4,27 @@
...
+    Math.sqrt(sample_variance)
   end
 
   def outliers(deviation)
-    m = self.mean
-    sd = self.standard_deviation
-    self.select { |n| (m-n).abs > (deviation * sd) }
+    m = mean
+    sd = standard_deviation
+    select { |n| (m-n).abs > (deviation * sd) }
```
Comment (alloy) [inline config/enumerable_stats.rb]: 😂

**Comment by orta** [comment]
OK - so now we need to think about gating - Danger is no use if she's commenting on every PR, ideas for how we can only show this when relevant? 

Maybe, only show then there are Swift file in modified?

**Comment by ashfurrow** [comment]
Makes sense. Would b nice to filter based on the specific swift files that were changed, if any.

**Comment by orta** [comment]
Hrm, but that would hide Pods that have the same issue. Though personally, I'm less worried about them as they're built once normally.

**Comment by mennenia** [comment]
what if, only comment if the outliners are a certain % over the average?, like, triple the average or something.

**Comment by orta** [comment]
> what if, only comment if the outliners are a certain % over the average?, like, triple the average or something.

This is what it currently does, it looks for compile times that are three times the standard deviation from everything else ( I believe that means things that fall outside of the 90% range roughly )

**Comment by orta** [comment]
Looks like that's outside the 99.7% average 
https://en.wikipedia.org/wiki/68–95–99.7_rule

**Comment by orta** [comment]
Hrm, I don't think I can think of a better way to do this, so let's do this as it is for the moment and we can see how it looks over some time.

**Comment by alloy** [comment]
👍 🎉

**Comment by KrauseFx** [comment]
Great addition, I love the things you build with `danger`. I wonder if we should already set up an `examples` repo, that just links to external `Dangerfile`s, so people get a sense of what you can do with it 👍

**Pull request merged**